utils/wxrc/wxrc.cpp

```diff
@@ -626,41 +627,105 @@ void XmlResApp::MakePackageZIP(const wxArrayString& flist)
 
-    const size_t lng = wx_truncate_cast(size_t, offset);
+    // [2024-03-15]
+    //
+    //   There are two environments (processes) to consider:
+    //     - the resource compiler;
+    //     - the process using the resource.
+    //
+    //   Previously, we used to check whether `offset` fitted into `size_t`.
+    //
+    //   But this used to check against the `size_t` of the resource compiler.
+    //   The check might have passed, but this might have been misleading
+    //   (if the resource compiler is a 64-bit process and the process using the resource is a 32-bit process).
+    //
+    //   Also, for Windows at least, the file format for modules (executables and shared libraries)
+    //   which are to be loaded in 64-bit processes
+    //   is not called PE64, but instead PE32+,
+    //   and yes, in this format virtual addresses are indeed 64-bit,
+    //   but in fact relative virtual addresses (and also file offsets) are 32-bit
+    //   (i.e. a single module cannot span for more than 4 GiB in virtual memory).
+    //
+    //   It follows that the check might have also been misleading
+    //   in the case of the resource compiler being a 64-bit Linux process
+    //   and the process using the resource being a Windows process -- even a 64-bit Windows process.
+    //
+    //   Therefore, we now check whether the size of the resource fits
+    //   not into `size_t`, but instead into `uint32_t`,
+    //   i.e. we do not support resources larger than 4 GiB.
+    //
+    const wxUint32 lng = wx_truncate_cast(wxUint32, offset);
     wxASSERT_MSG( static_cast<wxFileOffset>(lng) == offset,
                   wxT("Huge file not supported") );
 
     snum.Printf(wxT("%i"), num);
-    output.Printf(wxT("static size_t xml_res_size_") + snum + wxT(" = %lu;\n"),
-                  static_cast<unsigned long>(lng));
-    output += wxT("static unsigned char xml_res_file_") + snum + wxT("[] = {\n");
+
+    tmp.Printf(wxT("// File '%s':\n"), filename);
+    output += tmp;
+
+    tmp.Printf(wxT("const size_t xml_res_size_") + snum + wxT(" = %u;\n"), lng);
+    output += tmp;
+
+    output += wxT("const unsigned char xml_res_file_") + snum + wxT("[] = {\n");
     // we cannot use string literals because MSVC is dumb wannabe compiler
     // with arbitrary limitation to 2048 strings :(
 
-    unsigned char *buffer = new unsigned char[lng];
-    file.Read(buffer, lng);
+    std::vector<unsigned char> buffer(lng);
+    file.Read(buffer.data(), lng);
 
-    for (size_t i = 0, linelng = 0; i < lng; i++)
+    for (wxUint32 i = 0, linelng = 0, sectionlng = 0; ; )
     {
-        tmp.Printf(wxT("%i"), buffer[i]);
-        if (i != 0) output << wxT(',');
-        if (linelng > 70)
+        if (! i || i >= lng || sectionlng >= 0x100)
         {
-            linelng = 0;
-            output << wxT("\n");
+            if (linelng) wxFAIL;
+
+            tmp.Printf
+            (
+                wxT("// Offset %8Xh of %8Xh (%7.2lf%%)%c\n"),
```

Comment:
I think it might help with a quick vision of progress, telling us where we are within the file, for example when we scroll up and down (especially at fast speed) within the generated C++ code.

For me personally, I remember (and like) hex dumps in a format which also includes the offset:

```
$ hexdump -Cv "$(which ls)" | head -n 16
00000000  7f 45 4c 46 02 01 01 00  00 00 00 00 00 00 00 00  |.ELF............|
00000010  03 00 3e 00 01 00 00 00  60 61 00 00 00 00 00 00  |..>.....`a......|
00000020  40 00 00 00 00 00 00 00  68 37 02 00 00 00 00 00  |@.......h7......|
00000030  00 00 00 00 40 00 38 00  0b 00 40 00 1e 00 1d 00  |....@.8...@.....|
00000040  06 00 00 00 04 00 00 00  40 00 00 00 00 00 00 00  |........@.......|
00000050  40 00 00 00 00 00 00 00  40 00 00 00 00 00 00 00  |@.......@.......|
00000060  68 02 00 00 00 00 00 00  68 02 00 00 00 00 00 00  |h.......h.......|
00000070  08 00 00 00 00 00 00 00  03 00 00 00 04 00 00 00  |................|
00000080  a8 02 00 00 00 00 00 00  a8 02 00 00 00 00 00 00  |................|
00000090  a8 02 00 00 00 00 00 00  1c 00 00 00 00 00 00 00  |................|
000000a0  1c 00 00 00 00 00 00 00  01 00 00 00 00 00 00 00  |................|
000000b0  01 00 00 00 04 00 00 00  00 00 00 00 00 00 00 00  |................|
000000c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000d0  38 35 00 00 00 00 00 00  38 35 00 00 00 00 00 00  |85......85......|
000000e0  00 10 00 00 00 00 00 00  01 00 00 00 05 00 00 00  |................|
000000f0  00 40 00 00 00 00 00 00  00 40 00 00 00 00 00 00  |.@.......@......|
```

Including the offset on each line may be a little too much output (of C++ code), so I thought we just include it every 16 lines.

However, I understand that this is very personal and I can only ask that you consider it for a little time. If you prefer otherwise, an empty line or nothing at all, I am going to change the code accordingly.

(A separate commit has been created for this: https://github.com/wxWidgets/wxWidgets/pull/24437/commits/ec7d065c4e618db5c5718c864a232c4994f81072 -- as part of #24437. The exact commit ID's might change as that pull request is changed.)